native/dlaqr5.go

```diff
@@ -0,0 +1,616 @@
+// Copyright ©2016 The gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package native
+
+import (
+	"math"
+
+	"github.com/gonum/blas"
+	"github.com/gonum/blas/blas64"
+)
+
+// Dlaqr5 performs a single small-bulge multi-shift QR sweep on an isolated
+// block of a Hessenberg matrix.
+//
+// wantt and wantz determine whether the quasi-triangular Schur factor and the
+// orthogonal Schur factor, respectively, will be computed.
+//
+// kacc22 specifies the computation mode of far-from-diagonal orthogonal
+// updates. Permitted values are:
+//  0: Dlaqr5 will not accumulate reflections and will not use matrix-matrix
+//     multiply to update far-from-diagonal matrix entries.
+//  1: Dlaqr5 will accumulate reflections and use matrix-matrix multiply to
+//     update far-from-diagonal matrix entries.
+//  2: Dlaqr5 will accumulate reflections, use matrix-matrix multiply to update
+//     far-from-diagonal matrix entries, and take advantage of 2×2 block
+//     structure during matrix multiplies.
+// For other values of kacc2 Dlaqr5 will panic.
+//
+// n is the order of the Hessenberg matrix H.
+//
+// ktop and kbot are indices of the first and last row and column of an isolated
+// diagonal block upon which the QR sweep will be applied. It must hold that
+//  ktop == 0,   or 0 < ktop <= n-1 and H[ktop, ktop-1] == 0, and
+//  kbot == n-1, or 0 <= kbot < n-1 and H[kbot+1, kbot] == 0,
+// otherwise Dlaqr5 will panic.
+//
+// nshfts is the number of simultaneous shifts. It must be positive and even,
+// otherwise Dlaqr5 will panic.
+//
+// sr and si contain the real and imaginary parts, respectively, of the shifts
+// of origin that define the multi-shift QR sweep. On return both slices may be
+// reordered by Dlaqr5. Their length must be equal to nshfts, otherwise Dlaqr5
+// will panic.
+//
+// h and ldh represent the Hessenberg matrix H of size n×n. On return
+// multi-shift QR sweep with shifts sr+i*si has been applied to the isolated
+// diagonal block in rows and columns ktop through kbot, inclusive.
+//
+// iloz and ihiz specify the rows of Z to which transformations will be applied
+// if wantz is true. It must hold that 0 <= iloz <= ihiz < n, otherwise Dlaqr5
+// will panic.
+//
+// z and ldz represent the matrix Z of size n×n. If wantz is true, the QR sweep
+// orthogonal similarity transformation is accumulated into
+// z[iloz:ihiz,ilo:ihi] from the right, otherwise z not referenced.
+//
+// v and ldv represent an auxiliary matrix of size (nshfts/2)×3.
```

Comment:
I transposed it because of Dlaqr1. In fortran it takes columns of V but without a stride which is assumed to be 1. So I transpose and pass rows to Dlaqr1.
I could use the array `vt` to pass to Dlaqr1 everywhere and then copy its elements out as on L245 but it would be slightly less efficient. The question is whether it's worth it. Dlaqr5 is a very internal and specific routine that is called only from Dlaqr0 and Dlaqr4. They both pass a work slice for v and in fact at the call site nothing will change, ldv is 3 both in Fortran and here.
I will make the transpose more explicit in the comment but I think it's ok not to do anything else.